File: docs/source/src/rust/user-guide/expressions/folds.rs
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --8<-- [start:mansum]
    use polars::lazy::dsl::sum_horizontal;
    use polars::prelude::*;

    let df = df!(
        "label" => ["foo", "bar", "spam"],
        "a" => [1, 2, 3],
        "b" => [10, 20, 30],
    )?;

    let result = df
        .clone()
        .lazy()
        .select([
            fold_exprs(
                lit(0),
                PlanCallback::new(|(acc, val)| &acc + &val),
                [col("a"), col("b")],
                false,
                None,
            )
            .alias("sum_fold"),
            sum_horizontal([col("a"), col("b")], true)?.alias("sum_horz"),
        ])
        .collect()?;

    println!("{result:?}");
    // --8<-- [end:mansum]

    // --8<-- [start:mansum-explicit]
    let acc = lit(0);
    let f = |acc: Expr, val: Expr| acc + val;

    let result = df
        .clone()
        .lazy()
        .select([
            f(f(acc, col("a")), col("b")),
            fold_exprs(
                lit(0),
                PlanCallback::new(|(acc, val)| &acc + &val),
                [col("a"), col("b")],
                false,
                None,
            )
            .alias("sum_fold"),
        ])
        .collect()?;

    println!("{result:?}");
    // --8<-- [end:mansum-explicit]

    // --8<-- [start:manprod]
    let result = df
        .clone()
        .lazy()
        .select([fold_exprs(
            lit(0),
            PlanCallback::new(|(acc, val)| &acc * &val),
            [col("a"), col("b")],
            false,
            None,
        )
        .alias("prod")])
        .collect()?;

    println!("{result:?}");
    // --8<-- [end:manprod]

    // --8<-- [start:manprod-fixed]
    let result = df
        .lazy()
        .select([fold_exprs(
            lit(1),
            PlanCallback::new(|(acc, val)| &acc * &val),
            [col("a"), col("b")],
            false,
            None,
        )
        .alias("prod")])
        .collect()?;

    println!("{result:?}");
    // --8<-- [end:manprod-fixed]

    // --8<-- [start:conditional]
    let df = df!(
        "a" => [1, 2, 3],
        "b" => [0, 1, 2],
    )?;

    let result = df
        .lazy()
        .filter(fold_exprs(
            lit(true),
            PlanCallback::new(|(acc, val)| &acc & &val),
            [col("*").gt(1)],
            false,
            None,
        ))
        .collect()?;

    println!("{result:?}");
    // --8<-- [end:conditional]

    // --8<-- [start:string]
    let df = df!(
        "a" => ["a", "b", "c"],
        "b" => [1, 2, 3],
    )?;

    let result = df
        .lazy()
        .select([concat_str([col("a"), col("b")], "", false)])
        .collect()?;
    println!("{result:?}");
    // --8<-- [end:string]

    Ok(())
}

File: docs/source/src/python/user-guide/expressions/folds.py
# --8<-- [start:mansum]
import operator
import polars as pl

df = pl.DataFrame(
    {
        "label": ["foo", "bar", "spam"],
        "a": [1, 2, 3],
        "b": [10, 20, 30],
    }
)

result = df.select(
    pl.fold(
        acc=pl.lit(0),
        function=operator.add,
        exprs=pl.col("a", "b"),
    ).alias("sum_fold"),
    pl.sum_horizontal(pl.col("a", "b")).alias("sum_horz"),
)

print(result)
# --8<-- [end:mansum]

# --8<-- [start:mansum-explicit]
acc = pl.lit(0)
f = operator.add

result = df.select(
    f(f(acc, pl.col("a")), pl.col("b")),
    pl.fold(acc=acc, function=f, exprs=pl.col("a", "b")).alias("sum_fold"),
)

print(result)
# --8<-- [end:mansum-explicit]

# --8<-- [start:manprod]
result = df.select(
    pl.fold(
        acc=pl.lit(0),
        function=operator.mul,
        exprs=pl.col("a", "b"),
    ).alias("prod"),
)

print(result)
# --8<-- [end:manprod]

# --8<-- [start:manprod-fixed]
result = df.select(
    pl.fold(
        acc=pl.lit(1),
        function=operator.mul,
        exprs=pl.col("a", "b"),
    ).alias("prod"),
)

print(result)
# --8<-- [end:manprod-fixed]

# --8<-- [start:conditional]
df = pl.DataFrame(
    {
        "a": [1, 2, 3],
        "b": [0, 1, 2],
    }
)

result = df.filter(
    pl.fold(
        acc=pl.lit(True),
        function=lambda acc, x: acc & x,
        exprs=pl.all() > 1,
    )
)
print(result)
# --8<-- [end:conditional]

# --8<-- [start:string]
df = pl.DataFrame(
    {
        "a": ["a", "b", "c"],
        "b": [1, 2, 3],
    }
)

result = df.select(pl.concat_str(["a", "b"]))
print(result)
# --8<-- [end:string]
